src/contrib/tests/system-tests.rst

```diff
@@ -6,6 +6,164 @@ System Tests
+* Avoid cumbersome calls to the host and overly complicated test code.
+
+Add test coverage by parametrizing topologies and abstracting methods
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+* Increase test coverage by using multiple topologies and pre-defined topology groups. This makes the tests generic and compatible with multiple providers.
+* Extend the `SSSD Test Framework <https://tests.sssd.io/en/latest/>`_ and abstract methods to support multiple providers.
+
+An example of this is *Active Directory Group Policy Objects*. These tests require different calls to set up GPOs on AD and Samba. Abstracting the methods will extend the test to cover both topologies instead of two sets of tests.
+
+Guidelines
+------------
+
+All code must be fully typed and follow the black coding style. All code must be validated using the following tools:
+
+* Check PEP8 compliance with `flake8 <https://flake8.pycqa.org>`__ and `pycodestyle <https://pycodestyle.pycqa.org>`__.
+* Sort imports with `isort <https://pycqa.github.io/isort/>`__.
+* Convert to `black <https://black.readthedocs.io>`__ style.
+* Check types with `mypy <https://mypy.readthedocs.io>`__.
+
+Organizing test cases
+^^^^^^^^^^^^^^^^^^^^^
+
+Pytest allows you to write tests inside a class (starts with `Test`) or directly inside a module (a function
+starting with `test_`). Even though it might be logical to organize tests inside a class, it does not give you any
+benefit over plain function and it create just one more level of organization that must be correctly kept and
+maintained.
+
+.. warning::
+
+    **Avoid organizing tests into classes** *unless there is a good reason to use them* .
+
+File header
+^^^^^^^^^^^
+
+* The beginning of the file must contain the title and requirements.
+* The beginning of the file must provide a summary, not more than a couple of lines, detailing what the test covers.
+* A guide must be provided if the tests require additional knowledge to run; a great example is `passkey tests <https://github.com/SSSD/sssd/blob/master/src/tests/system/tests/test_passkey.py>`__.
+
+Naming
+^^^^^^^
+
+Test names contain three parts, the **pytest discovery pattern**, **file name**, and **test name**.
+
+* **test**\_ Pytest discovery pattern.
+* test\_ **filename**\__ File name, no extension and a double underscore separating the file name from the test name.
+* test_filename\__ **name_of_the_test** Test name (see below).
+
+Naming criteria
+'''''''''''''''''
+
+* Test names should describe the expected behavior of what you are testing.
+* Test names should describe what the code is doing.
+* Test names should provide a good idea of what the test does.
+* Test names can use the filename to help describe what the test does.
+
+Docstring
+'''''''''''
+
+All tests require the following documentation strings.
+
+* **name**: Test name without the_under_scores.
```

Comment:
Could it be 'Title', since we use title everywhere?"

Reply:
Title, is the wrong word to use here. It's used to distinguish the work. Like a book title, title of a play. We can make it more descriptive, like **test name**, **test case**. Where have you seen test scenarios called titles?